presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -570,6 +575,12 @@ public PlanOptimizers(
                         .add(new InlineProjections(metadata.getFunctionAndTypeManager()))
                         .build()));
 
+        // We do a single pass, and assign canonical ids to plan nodes, and precompute their hashes.
+        // These hashes are only computed once during planning.
+        // After this step, nodes with same canonical plan id will share same hashes, and thus same history based statistics.
+        // TODO: Ensure rules after this preserve canonical plan ids when needed.
+        builder.add(new StatsRecordingPlanOptimizer(optimizerStats, new HistoricalStatisticsRecordingPlanOptimizer(objectMapper)));
```

Comment:
I would give HistoricalStatisticsRecordingPlanOptimizer a different name to avoid confusion with StatsRecordingPlanOptimizer since they are quite different from each other.

StatsRecordingPlanOptimizer is just a wrapper around another plan optimizer that logs how long that optimizer took to run

HistoricalStatsRecordingPlanOptimizer adds the cannonicalId to the plan and registers plan hashes (and doesn't record stats at all)